### PR TITLE
feat: highlight nav item in onPageNav ToC

### PR DESCRIPTION
**Pull request by tusharf5**
## Motivation

Highlight nav item in onPageNav ToC when scrolling  #1024

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Local Testing.

**Before Changes** - https://www.loom.com/share/956ba8b3696a41ba841d38fa5197ab20

**After Changes** - https://www.loom.com/share/68b57bf4af47400ab9abbdd1b8cbd3f6

## Related PRs

[#1024](https://github.com/facebook/Docusaurus/issues/1024)

### CODE REVIEW

File: packages/docusaurus-1.x/lib/core/Head.js [hunk 1]
```diff
@@ -172,6 +172,35 @@ class Head extends React.Component {
             }}
           />
         )}
+        <script src="https://unpkg.com/smooth-scroll@16.0.3/dist/smooth-scroll.polyfills.min.js" />
+        <script src="/js/scrollSpy.js" />
+        <script
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/core/Head.js]: Why don't put the contents of this script into `scrollSpy.js`? We'd wanna avoid using strings where possible as they're harder to statically analyze.

(This is assuming we want to keep this scroll on click feature which I'm not in favor of)

File: packages/docusaurus-1.x/lib/core/Head.js [hunk 2]
```diff
@@ -172,6 +172,35 @@ class Head extends React.Component {
...
+          dangerouslySetInnerHTML={{
+            __html: `
+              document.addEventListener('DOMContentLoaded', () => {
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/core/Head.js]: Use the ES5 function syntax because these JS scripts/code will not be transpiled at all.

File: packages/docusaurus-1.x/lib/static/css/main.css [hunk 3]
```diff
@@ -1975,6 +1975,10 @@ input::placeholder {
   color: #717171;
 }
 
+.onPageNav .toc-headings>li>a.active {
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/css/main.css]: Please follow the coding style in the rest of this CSS file in terms of spacing and indentation.

File: packages/docusaurus-1.x/lib/static/css/main.css [hunk 4]
```diff
@@ -1975,6 +1975,10 @@ input::placeholder {
...
+    font-weight: 700!important;
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/css/main.css]: Why is `!important` needed? Adding `!important` makes it hard to override by users who don't want this behavior.

Comment (yangshun) [reply packages/docusaurus-1.x/lib/static/css/main.css]: Use 600 as it's the more common value in the code.

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+          const top = nextHeader.getBoundingClientRect().top;
+          // The following tests whether top + scrollTop
+          // (the top of the header) is greater than scrollTop
+          // (where scrollTop = window.pageYOffset, the top of
+          // the window), with OFFSET pixels of slop.
+          current = top > OFFSET;
+        }
+        if (current) {
+          found = true;
+          headings[i].className = 'active';
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/js/scrollSpy.js]:
```js
headings[i].classList.add('active');
```

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        } else {
+          headings[i].className = '';
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: This erases all class names on the element and will break if we add more classes in future.

```js
headings[i].classList.remove('active');
```

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+(function scrollSpy() {
+  const OFFSET = 10;
+  let timer;
+  let headingsCache;
+  const findHeadings = () =>
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: Use `function` declarations for this entire file (no arrow functions).

**Comment by endiliey** [comment]
Sorry but I’m not in favor of this PR.

First, it can be easily added by the user as a script. Katex did it (i can argue that the code is similar too) https://github.com/KaTeX/KaTeX/blob/master/website/static/js/scrollspy.js

Second, we are working on v2 and this means we have to port it back to v2. With such limited manpower (nobody is working on docusaurus full-time), its just going to delay v2 even more.

Third, new feature can come with a new bug. And eventually if bug arise, the maintainer will be the one having to fix it.

I think we need to make a clear stance that we are less likely to accept new feature as we are working on v2. Unless there is a high demand on it.

I really appreciate you taking your time to send this PR. I wish the effort was sent towards v2 instead

**Comment by wgao19** [comment]
I like this addition! Feels very natural to have the highlighting follow scrolling.

I agree with @yangshun in favor of the browser's default behavior. I'd say smooth scroll on header clicks is at most an opt-in. 

Checked a few non-Docusaurus doc sites on this issue:

|Doc site|Highlight follows scroll|Smooth scroll to header clicks|
|--------|-|-|
|[ReactJS](https://reactjs.org/docs/hello-world.html)|No|No|
|[VueJS](https://vuejs.org/v2/api/)|Yes|Yes|
|[Docz](https://www.docz.site/docs/)|Yes|No|
|[Google Web](https://developers.google.com/web/updates/2016/02/smooth-scrolling-in-chrome-49)|Yes|No|

**Comment by tusharf5** [comment]
Thank you all for the great feedback. @endiliey If this PR gets merged then I can certainly work on sending a PR that replicates this behavior in v2 also.

**Comment by endiliey** [comment]
I'm going to let yall decide. My humble request and only complain is that we have to port this to v2 first before merging it. Not a good idea to see it in v1 but not in v2.

This was unexpected as well because there seems to be no discussion regarding this and suddenly the PR came out

**Comment by endiliey** [comment]
> Thank you all for the great feedback. @endiliey If this PR gets merged then I can certainly work on sending a PR that replicates this behavior in v2 also.

Oh cool, that'll be awesome then. Then my only complain is pretty much gone

**Comment by tusharf5** [comment]
@yangshun I've removed smooth scroll from this PR. Let me know if a separate PR for that would make sense : )

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      bounding.top >= 0 &&
+      bounding.bottom <=
+        (window.innerHeight || document.documentElement.clientHeight)
+    ) {
+      // Already visible.  Do nothing.
+    } else {
+      // Not visible.  Scroll sidebar.
+      item.scrollIntoView({block: 'center', inline: 'nearest'});
+      // eslint-disable-next-line no-multi-assign
+      document.body.scrollTop = document.documentElement.scrollTop = 0;
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: Can you explain what these are for?

File: packages/docusaurus-1.x/lib/static/css/main.css [hunk 9]
```diff
@@ -725,7 +725,7 @@ blockquote {
   position: fixed;
   width: 100%;
   z-index: 9999;
-  transform: translate3d(0,0,0);
+  transform: translate3d(0, 0, 0);
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/css/main.css]: Thanks for fixing all these style inconsistencies, but we'd prefer to keep them out of this PR as it makes it harder to review (could make another PR to address styling issues).

File: packages/docusaurus-1.x/lib/core/Head.js [hunk 10]
```diff
@@ -172,6 +172,7 @@ class Head extends React.Component {
             }}
           />
         )}
+        <script src="/js/scrollSpy.js" />
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/core/Head.js]: This assumes a baseUrl of /, which is an incorrect assumption. See L186 where a similar script was added for the codetabs functionality.

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    // Cache the headings once the page has fully loaded.
+    headingsCache = findHeadings();
+    onScroll();
+    // Find the active nav item in the sidebar
+    const item = document.getElementsByClassName('navListItemActive')[0];
+    if (!item) {
+      return;
+    }
+    const bounding = item.getBoundingClientRect();
+    if (
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: I don't really understand what this entire block does, do you mind explaining?

Also it's really uncommon to have an empty `if` block. It'd be better to negate the condition and shift the `else` into the negated `if` condition.

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js (hunk 8 above)
Comment (tusharf5) [reply packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: I realized these lines were unnecessary so I've removed them.

File: packages/docusaurus-1.x/lib/static/css/main.css (hunk 9 above)
Comment (tusharf5) [reply packages/docusaurus-1.x/lib/static/css/main.css]: Removed this and one more.

File: packages/docusaurus-1.x/lib/core/Head.js (hunk 10 above)
Comment (tusharf5) [reply packages/docusaurus-1.x/lib/core/Head.js]: Done

File: packages/docusaurus-1.x/lib/static/js/scrollSpy.js (hunk 11 above)
Comment (tusharf5) [reply packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: Removed unnecessary code from here and also I've left a lot of comments and modified variable names to better understand what is going on. If it still doesn't make any sense I will be happy to explain it here.

Comment (wgao19) [reply packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: I'm also trying to learn the chunk, here's my understanding, correct me if I'm wrong, and can consider the following comment?

```javascript
(function scrollSpy() {
  const OFFSET = 10;
  let timer;
  let headingsCache;
  const findHeadings = function findHeadings() {
    return headingsCache || document.querySelectorAll('.toc-headings > li > a');
  };
  const onScroll = function onScroll() {
    if (timer) {
      // throttle
      return;
    }
    /**
     * On every call, try to find header right after  <-- next header
     * the one whose content is on the current screen <-- highlight this
     */
    timer = setTimeout(function() {
      timer = null;
      let activeNavFound = false;
      const headings = findHeadings(); // all of toc nav anchors
      for (let i = 0; i < headings.length; i++) {
        let currNavActive = !activeNavFound;
        /**
         * Enter the following check up only when an active nav header is not yet found
         * Then, check the bounding rectangle of the next header
         * The headers that are scrolled passed will have negative bounding rect top
         * So the first one with positive bounding rect top will be the nearest next header
         */
        if (currNavActive && i < headings.length - 1) {
          const next = headings[i + 1].href.split('#')[1];
          const nextHeader = document.getElementById(next);
          const top = nextHeader.getBoundingClientRect().top;
          currNavActive = top > OFFSET;
        }
        /**
         * Stop searching once a first such header is found,
         * this makes sure the highlighted header is the most current one
         */
        if (currNavActive) {
          activeNavFound = true;
          headings[i].classList.add('active');
        } else {
          headings[i].classList.remove('active');
        }
      }
    }, 100);
  };
```

Comment (tusharf5) [reply packages/docusaurus-1.x/lib/static/js/scrollSpy.js]: Yupp that's correct. I will update the comments. Thank you :)

**Comment by tusharf5** [comment]
Working on it. Will submit a PR soon 🙋‍♂️
